packages/eslint-plugin-baseui/src/deprecated-theme-api.js

```diff
@@ -0,0 +1,434 @@
+              );
+            };
+          }
+
+          // Check if we are setting a property in createTheme:overrides
+          // Ex: createTheme({}, {datepickerBackground: "pink"});
+          if (
+            context
+              .getAncestors()
+              .some(
+                node =>
+                  node.type === 'CallExpression' &&
+                  node.callee.name === 'createTheme',
+              ) &&
+            node.parent.type === 'Property' &&
+            node === node.parent.key
+          ) {
+            // A deprecated property is being set in createTheme.
+            // Assuming a violation.
+            context.report({
+              node,
+              messageId,
+              data,
+              fix,
+            });
+            return;
+          }
+
+          // Exit early if we are not in a function.
+          // We don't have access to the theme outside of functions.
+          // Ex: styled, withStyle, overrides, function component (useStyletron)
+          if (scope.type !== 'function') {
+            return;
+          }
+
+          // Exit early if the identifier is a key in an object expression
+          // Ex: { background: $theme.colors.white }
+          if (
+            node.parent.type === 'Property' &&
+            node.parent.parent.type === 'ObjectExpression' &&
+            node === node.parent.key
+          ) {
+            return;
+          }
+
+          // Exit early if identifier is a value in an object destructuing pattern
+          // Ex: ({$theme: {colors: {background}}}) => {}
+          //  - background will match twice, as both a key and as a value, so we skip the value match
+          if (
+            node.parent.type === 'Property' &&
+            node.parent.parent.type === 'ObjectPattern' &&
+            node === node.parent.value
+          ) {
+            return;
+          }
+
+          // Check if our identifier is accessing properties on a theme, $theme, or its theme concern.
+          // Ex: theme.colors.foreground
+          // Ex: props.$theme.colors.foreground
+          // Ex: colors.foreground
+          if (node.parent.type === 'MemberExpression') {
+            const expressionText = sourceCode.getText(node.parent);
+            if (
+              expressionText.search('theme.') ||
+              expressionText.search(
+                deprecatedThemeProperties[identifier].concern,
+              )
+            ) {
+              // Expression accesses properties on a theme, $theme, or colors object.
+              // Assuming a violation.
+              context.report({
+                node,
+                messageId,
+                data,
+                fix,
+              });
+              return;
+            }
+          }
+
+          // Check if our identifier was fully destructured from a $theme object.
+          // Ex: styled('div', ({$theme: {colors: {background}}}) => {})
+          scope.block.params.forEach(param => {
```

Comment:
can this be checked on the function name matching `styled` or `withStyle`?

Reply:
I think one nice thing with this current approach is that it will catch functions that are defined before being passed to `styled|withStyle|overrides` because it keys on `({$theme})` in the parameters of the function. I figured this type of de-structuring signature is unique enough to our styling utilities that it didn't require further refining.

Reply:
I personally feel that this should 'miss' sometimes rather than be overzealous in flagging errors

Reply:
In refactor only functions passed to style, withStyle, and overrides style properties will have a chance to flag an error.

Reply:
sweeeet